src/test/librbd/fsx.cc

```diff
@@ -1343,6 +1353,294 @@ const struct rbd_operations nbd_operations = {
+  process2.add_cmd_arg("show");
+  process2.add_cmd_arg(img.c_str());
+
+  r = process2.spawn();
+  if (r < 0) {
+    prt("rbd-wnbd show, error: %s\n", process2.err().c_str());
+    return r;
+  }
+
+  // This should be more than enough for the "rbd-wnbd show" output
+  char map_info[4096];
+  r = safe_read(process2.get_stdout(), map_info, sizeof(map_info));
+  if (r < 0) {
+    prt("unable to retrieve rbd-wnbd map info, error: %d\n", r);
+    return r;
+  }
+
+  r = process2.join();
+  if (r) {
+    prt("rbd-wnbd show failed, error: %s\n", process2.err().c_str());
+    return -EINVAL;
+  }
+
+  std::string dev_path;
+  std::regex dev_num_re("\"disk_number\": ([0-9]+)");
+  std::smatch cm;
+  std::string map_info_str(map_info);
+  if (std::regex_search(map_info_str, cm, dev_num_re)) {
+    dev_path = std::string("\\\\.\\PhysicalDrive") + cm.str(1);
+  } else {
+    prt("unable to retrieve rbd-wnbd device number, device info: %s\n",
+      map_info);
+    return -EINVAL;
+  }
+
+  prt("rbd-wnbd: mapped image %s/%s as: %s\n",
+    pool, name, dev_path.c_str());
+
+  fd = open(dev_path.c_str(), O_RDWR | O_BINARY | o_direct);
+  if (fd < 0) {
+    r = -errno;
+    prt("open(%s) failed\n", dev_path.c_str());
+    return r;
+  }
+
+  ctx->krbd_name = strdup(img.c_str());
+  ctx->krbd_fd = fd;
+
+  return wnbd_set_writable(ctx);
+}
+
+int
+wnbd_close(struct rbd_ctx *ctx)
+{
+  int r;
+
+  ceph_assert(ctx->krbd_name && ctx->krbd_fd >= 0);
+
+  if (close(ctx->krbd_fd) < 0) {
+    r = -errno;
+    prt("close(%s) failed\n", ctx->krbd_name);
+    return r;
+  }
+
+  SubProcess process("rbd-wnbd");
+  process.add_cmd_arg("unmap");
+  process.add_cmd_arg(ctx->krbd_name);
+
+  r = process.spawn();
+  if (r < 0) {
+    prt("rbd-wnbd unmap failed, error: %s\n", process.err().c_str());
+    return r;
+  }
+  r = process.join();
+  if (r) {
+    prt("rbd-wnbd unmap failed, error", process.err().c_str());
+    return -EINVAL;
+  }
+
+  free((void *)ctx->krbd_name);
+
+  ctx->krbd_name = NULL;
+  ctx->krbd_fd = -1;
+
+  return __librbd_close(ctx);
+}
+
+ssize_t
+wnbd_read(struct rbd_ctx *ctx, uint64_t off, size_t len, char *buf)
+{
+  ssize_t n;
+
+  n = pread(ctx->krbd_fd, buf, len, off);
+  if (n < 0) {
+    n = -errno;
+    prt("pread(%llu, %zu) failed\n", off, len);
+    return n;
+  }
+
+  return n;
+}
+
+
+ssize_t
+wnbd_write(struct rbd_ctx *ctx, uint64_t off, size_t len, const char *buf)
+{
+  ssize_t n;
+
+  n = pwrite(ctx->krbd_fd, buf, len, off);
+  if (n < 0) {
+    n = -errno;
+    prt("pwrite(%llu, %zu) failed\n", off, len);
+    return n;
+  }
+
+  return n;
+}
+
+
+int
+wnbd_flush(struct rbd_ctx *ctx)
+{
+  return _commit(ctx->krbd_fd);
+}
+
+int
+wnbd_discard(struct rbd_ctx *ctx, uint64_t off, uint64_t len)
+{
+  // TODO: use one of the following:
+  // * IOCTL_SCSI_PASS_THROUGH
+  // * IOCTL_SCSI_PASS_THROUGH_DIRECT
+  // * FSCTL_FILE_LEVEL_TRIM
+  // * sg3_utils
+  return librbd_discard(ctx, off, len);
```

Comment:
Not following how this would work -- a range would be zeroed in the image but not in the page cache (or its equivalent on Windows)?  If the idea is that fsx would be always run with `-H` passed to disable discards, then I think it's better to just omit the handler rather than provide a broken one.